pnpm-script-extended/pnpm-shell-completion-extended.plugin.zsh

```diff
@@ -0,0 +1,242 @@
+        
+        # Use _describe to present the completions
+        if [[ ${#completions[@]} -gt 0 ]]; then
+            _pnpm_tab_debug "Found ${#completions[@]} completions, calling _describe"
+            _describe "completions" completions
+            return 0
+        fi
+    fi
+    
+    _pnpm_tab_debug "No completions found, falling back to file completion"
+    # If we couldn't get or process completions, fall back to file completion
+    _files
+    return 1
+}
+
+# Check if a CLI supports __complete by running it through pnpm
+_pnpm_cli_has_completions() {
+    local cli_name=$1
+    _pnpm_tab_debug "Checking if $cli_name has completions via pnpm"
+    
+    # Try to execute the __complete command through pnpm
+    if output=$(_pnpm_run_cli $cli_name "__complete" 2>/dev/null) && [[ -n "$output" ]]; then
+        _pnpm_tab_debug "$cli_name supports completions via pnpm: $output"
+        return 0
+    fi
+    
+    _pnpm_tab_debug "$cli_name does not support completions via pnpm"
+    return 1
+}
+
+# Original pnpm-shell-completion logic
+if command -v pnpm-shell-completion &> /dev/null; then
+    pnpm_comp_bin="$(which pnpm-shell-completion)"
+    _pnpm_tab_debug "Found pnpm-shell-completion at $pnpm_comp_bin"
+else
+    pnpm_comp_bin="$(dirname $0)/pnpm-shell-completion"
+    _pnpm_tab_debug "Using relative pnpm-shell-completion at $pnpm_comp_bin"
+fi
+
+_pnpm() {
+    typeset -A opt_args
+    _pnpm_tab_debug "Starting pnpm completion, words: ${words[*]}"
+
+    _arguments \
+        '(--filter -F)'{--filter,-F}'=:flag:->filter' \
+        ':command:->scripts' \
+        '*:: :->command_args'
+
+    local target_pkg=${opt_args[--filter]:-$opt_args[-F]}
+    _pnpm_tab_debug "State: $state, target_pkg: $target_pkg"
+        
+    case $state in
+        filter)
+            if [[ -f ./pnpm-workspace.yaml ]] && [[ -x "$pnpm_comp_bin" ]]; then
+                _pnpm_tab_debug "Using pnpm-shell-completion for filter packages"
+                _values 'filter packages' $(FEATURE=filter $pnpm_comp_bin)
+            else
+                _pnpm_tab_debug "No workspace or pnpm-shell-completion for filter"
+                _message "package filter"
+            fi
+            ;;
+        scripts)
+            if [[ -x "$pnpm_comp_bin" ]]; then
+                _pnpm_tab_debug "Using pnpm-shell-completion for scripts"
+                _values 'scripts' $(FEATURE=scripts TARGET_PKG=$target_pkg ZSH=true $pnpm_comp_bin) \
+                    add remove install update publish
+            else
+                _pnpm_tab_debug "Using basic pnpm commands (no pnpm-shell-completion)"
+                _values 'scripts' \
+                    'add:Add a dependency' \
+                    'install:Install dependencies' \
+                    'remove:Remove a dependency' \
+                    'update:Update dependencies' \
+                    'publish:Publish package' \
+                    'run:Run script'
+            fi
+            ;;
```

Comment:
i'm just throwing an idea here, but dunno if you explored the possibility of having a caching utility and the use it within the filter completion case to use the cache and then finally clean that cache file when the script exits.